src/__tests__/pinning.test.js

```diff
@@ -1,398 +1,342 @@
+const TestClient = require('./testClient')
+const { registerMock3idResolver } = require('./mock3id')
+
+// Needed for ipfs spinup/teardown
+jest.setTimeout(15000)
+
+const pinningIpfsConfig = {
+  Bootstrap: [],
+  Addresses: {
+    Swarm: [
+      '/ip4/127.0.0.1/tcp/4002',
+      '/ip4/127.0.0.1/tcp/4003/ws'
+    ]
+  }
+}
 
-// const PRIV_IMG_ONLY_VALUES = {
-//   quiet: PRIVATE_QUIET.value,
-//   shh: PRIVATE_SHH.value
-// }
+const analyticsMock = {
+  trackPinDB: jest.fn(),
+  trackSyncDB: jest.fn(),
+  trackSpaceUpdate: jest.fn(),
+  trackPublicUpdate: jest.fn(),
+  trackRootUpdate: jest.fn(),
+  trackThreadUpdate: jest.fn(),
+  trackPrivateUpdate: jest.fn(),
+  trackPinDBAddress: jest.fn(),
+  trackSpaceUpdateByApp: jest.fn()
+}
 
-const THREEID_MOCK = {
-  DID: 'did:3:asdfasdf',
-  getKeyringBySpaceName: () => {
-    return {
-      getPublicKeys: () => {
-        return { signingKey: '044f5c08e2150b618264c4794d99a22238bf60f1133a7f563e74fcf55ddb16748159872687a613545c65567d2b7a4d4e3ac03763e1d9a5fcfe512a371faa48a781' }
-      }
-    }
+const mockProfileData = {
+  public: {
+    name: { timeStamp: 12000, value: 'very name' },
+    image: { timeStamp: 13000, value: 'such picture' }
   },
-  signJWT: payload => {
-    return didJWT.createJWT(payload, {
-      signer: didJWT.SimpleSigner('95838ece1ac686bde68823b21ce9f564bc536eebb9c3500fa6da81f17086a6be'),
-      issuer: 'did:3:asdfasdf'
-    })
+  private: {
+    shh: { timeStamp: 14000, value: 'many secret' },
+    quiet: { timeStamp: 15000, value: 'wow!' }
   }
 }
-// we need to have a fake 3id resolver since we have a fake 3id
-const register3idResolver = () => registerMethod('3', async () => {
-  return {
-    '@context': 'https://w3id.org/did/v1',
-    id: 'did:3:asdfasdf',
-    publicKey: [{
-      id: 'did:3:asdfasdf#signingKey',
-      type: 'Secp256k1VerificationKey2018',
-      publicKeyHex: '044f5c08e2150b618264c4794d99a22238bf60f1133a7f563e74fcf55ddb16748159872687a613545c65567d2b7a4d4e3ac03763e1d9a5fcfe512a371faa48a781'
-    }],
-    authentication: [{
-      type: 'Secp256k1SignatureAuthentication2018',
-      publicKey: 'did:3:asdfasdf#signingKey'
-    }]
-  }
-})
 
-const cache = {
-  write: jest.fn()
+const mockThreadEntries = [
+  { message: 'a great post' },
+  { message: 'another great post' }
+]
+
+function addReplicatedEmitter (pinning) {
+  pinning.events = new EventEmitter()
+  const origOpenDB = pinning.openDB
+  function myOpenDB (address, responseFn, onReplicatedFn, rootStoreAddress, analyticsFn) {
+    const newReplicatedFn = (odbAddress) => {
+      const numEntries = pinning.openDBs[odbAddress].db._oplog.values.length
+      pinning.events.emit('replicated', { odbAddress, numEntries })
+      if (onReplicatedFn) {
+        onReplicatedFn(odbAddress)
+      }
+    }
+    origOpenDB.call(pinning, address, responseFn, newReplicatedFn, rootStoreAddress, analyticsFn)
+  }
+  pinning.openDB = myOpenDB
+  return pinning
```

Comment:
The need for this function makes it quite clear that we should try to change the pinning-node code to be more event driven in general. For someone that's new to the code base I think this is super difficult to understand what it does.

Reply:
+1 for more event driven in future

Reply:
but this looks good 👍